Bottom: 67cef88dce9ad87900c7bbe461e14ca01e280f09
Top:    b967c99a1264a3953e5343a78eb32be3ab16c0b9
Author: Parth Shah <parth@linux.ibm.com>
Date:   2020-05-04 15:26:40 +0530

sched/idle: Add debugging bits to validate inconsistency in latency
sensitive task calculations

We monitor the task entering/exiting the scheduler, but there might be
unhandled situations which can lead to inconsistent value of the
nr_lat_sensitive counter. This may lead to restricting the use of IDLE
states despite absence of any latency sensitive workload.
Hence, add WARN_ON() if a negative value of nr_lat_sensitive value is found.

Signed-off-by: Parth Shah <parth@linux.ibm.com>


---

diff --git a/kernel/sched/idle.c b/kernel/sched/idle.c
index b32631292271..6581613189d9 100644
--- a/kernel/sched/idle.c
+++ b/kernel/sched/idle.c
@@ -231,6 +231,10 @@ static void cpuidle_idle_call(void)
 static void do_idle(void)
 {
 	int cpu = smp_processor_id();
+	int pm_disabled = atomic_read(&per_cpu(nr_lat_sensitive, cpu));
+
+	if (pm_disabled < 0)
+		pr_info("Inconsistent value of nr_lat_sensitive counter\n");
 
 	/*
 	 * If the arch has a polling bit, we maintain an invariant:
@@ -264,7 +268,7 @@ static void do_idle(void)
 		 * idle as we know that the IPI is going to arrive right away.
 		 */
 		if (cpu_idle_force_poll || tick_check_broadcast_expired() ||
-		    atomic_read(&per_cpu(nr_lat_sensitive, cpu))) {
+		    pm_disabled) {
 			tick_nohz_idle_restart_tick();
 			cpu_idle_poll();
 		} else {
